pkg/git/git.go

```diff
@@ -84,53 +86,92 @@ func Fetch(logger *zap.SugaredLogger, spec FetchSpec) error {
 		return err
 	}
 
-	fetchArgs := []string{"fetch", "--recurse-submodules=yes"}
+	fetchArgs := []string{"fetch"}
+	if spec.Submodules {
+		fetchArgs = append(fetchArgs, "--recurse-submodules=yes")
+	}
 	if spec.Depth > 0 {
 		fetchArgs = append(fetchArgs, fmt.Sprintf("--depth=%d", spec.Depth))
 	}
-	fetchArgs = append(fetchArgs, "origin", spec.Revision)
 
-	if _, err := run(logger, "", fetchArgs...); err != nil {
-		// Fetch can fail if an old commitid was used so try git pull, performing regardless of error
-		// as no guarantee that the same error is returned by all git servers gitlab, github etc...
-		if _, err := run(logger, "", "pull", "--recurse-submodules=yes", "origin"); err != nil {
-			logger.Warnf("Failed to pull origin : %s", err)
-		}
-		if _, err := run(logger, "", "checkout", spec.Revision); err != nil {
+	// Fetch the revision and verify with FETCH_HEAD
+	fetchParam := []string{spec.Revision}
+	checkoutParam := "FETCH_HEAD"
+
+	if spec.Refspec != "" {
+		// if refspec is specified, fetch the refspec and verify with provided revision
+		fetchParam = strings.Split(spec.Refspec, " ")
+		checkoutParam = spec.Revision
+	}
+
+	fetchArgs = append(fetchArgs, "origin", "--update-head-ok")
```

Comment:
I didn't recognize this flag and looked it up in the docs, which say:

> By default git fetch refuses to update the head which corresponds to the current branch. This flag disables the check. This is purely for the internal use for git pull to communicate with git fetch, and unless you are implementing your own Porcelain you are not supposed to use it.

I'm not sure what Porcelain is in this context but it would be great to include a note here on what the flag does and why we need to use it given that it's mostly intended for internal-to-git usage.

Reply:
I've added a comment w/ explanation. Required because git init checks out the master branch by default.